<img src="https://cdn-icons-png.flaticon.com/512/10310/10310245.png" alt="Logo" width="60" align="left" style="margin-right: 10px;"/>

# Calculator

## 0. About

To demonstrate functionality of the Calculator class, these cells execute the different arithmetic operations and methods in the packaged module.
The calculator can perform:
- Addition / Subtraction
- Multiplication / Division
- Take (n) root of a number
- Reset memory

This project was developed by Christine Angela Garcia in line with Sprint 2 Project 2 specifications provided by Turing College's Data Science program.

## 1. Import
To use the Calculator we must import the main class file from its module, found in the same directory as this notebook.

In [92]:
from calculator.calculator import Calculator

## 2. Functionality

### 2.0 New Calculator object

We can create a new Calculator object and verify its supposed starting memory: 0

In [93]:
calc = Calculator()
print(calc.memory)

0


The Calculator class also includes a *precision* attribute - setting the number of decimal places shown after the decimal. This helps prevent floating point number errors and inaccuracies. Its default value is 4.

To get an overview of the object and its properties we could print the object itself and view the following default values:
- memory = 0
- precision = 4

In [94]:
print(calc)

memory: 0, precision: 4


### Methods and Operations

#### 2.1. Addition & Subtraction

The *add(num)* method will add value *num* to the calculator's current memory and subtract(num) will subtract value *num*:

In [95]:
calc.add(22)
calc.add(6)
calc.add(4.2) 
calc.add(-2)

calc.subtract(10)
calc.subtract(2.55)
calc.subtract(-2)

Performing the above operations should yield a result of 19.65. However, the current value in memory will not be returned until calling:

#### 2.2. Equals

The calculator's value is stored internally until calling upon it with the *equals()* method -- this design emulates the press of an equals ('=') button on a calculator.

In [96]:
calc.equals()

19.65

#### 2.3. Multiplication & Division

Similar to the previous arithmetic methods, *multiply(num)* and *divide(num)* will perform the operations respectively.

In [97]:
calc.multiply(16) 
calc.divide(4)
calc.equals()

78.6

Resulting in 78.6

#### 2.4. Take (n) root of a number

The instance method *root(n)*  will take the n-th root of the *memory* value.

In [98]:
calc.root(3)
calc.equals()

4.2836

#### 2.5 Set Precision

Adjusting the precision property of the calculator is also an option, 

For example, we can adjust it to set 2 decimal places instead of the current 4 as so:

In [99]:
calc.precision = 2
calc.equals()

4.28

#### 2.6 Reset memory

So far, the calculator has been keeping a running memory, performing actions using the value stored in its memory property. At this point, the value stored is *4.2836*.

The *reset()* method will clear the calculator's memory, returning the value to 0.

In [100]:
calc.reset()

0

From here, methods will perform operations from 0 and not any previous number.

In [101]:
calc.add(10)
calc.divide(5)
calc.subtract(12)
calc.equals()

-10.0

### Edge Cases and Mathematical Impossibilities

Within the calculator class (specifically in the divide and root methods), there are checks in place to prevent mathematical impossiblities, such as:
1. Dividing by zero:

In [102]:
try: 
    calc.divide(0)
except ValueError as e:
    print(e)

Cannot divide by 0


2. Attempting to take the 0-th root of a number:

In [103]:
try: 
    calc.root(0)
except ValueError as e:
    print(e)

The root 'n' cannot be 0


3. Getting the even-numbered root of a negative number:

In [104]:
try: 
    calc.root(2)
except ValueError as e:
    print(e)

Cannot take even root of a negative number


In [105]:
calc.reset()

0

### Free Demo Cell

The following cell is meant as a demo space to test another instance of Calculator. Feel free to test any of the previously shown methods below:

In [106]:
new_calc = Calculator()
new_calc.add(10)
# Code here


new_calc.equals()

10

## Conclusion

Overall, having an object-oriented calculator class allows us to create an instance of it with its methods (root, reset, etc.) readily available. 

Thank you for viewing this notebook and its accompanying code. 